fido2.go

```diff
@@ -234,56 +238,101 @@ func DeviceLocations() ([]*DeviceLocation, error) {
 }
 
 // NewDevice opens device at path.
+// Opened devices must be explicitly closed.
 func NewDevice(path string) (*Device, error) {
 	if path == "" {
 		return nil, errors.Errorf("empty device path")
 	}
+
+	dev, err := openDevice(path)
+	if err != nil {
+		return nil, fmt.Errorf("open device: %w", err)
+	}
+
 	return &Device{
-		path: fmt.Sprintf("%s", path),
+		path: path,
+		dev:  dev,
 	}, nil
 }
 
-func (d *Device) open() (*C.fido_dev_t, error) {
+func openDevice(path string) (*C.fido_dev_t, error) {
+	pathC := C.CString(path)
+	defer C.free(unsafe.Pointer(pathC))
+
 	dev := C.fido_dev_new()
-	if cErr := C.fido_dev_open(dev, C.CString(d.path)); cErr != C.FIDO_OK {
-		return nil, errors.Wrap(errFromCode(cErr), "failed to open")
+	if cErr := C.fido_dev_open(dev, pathC); cErr != C.FIDO_OK {
+		return nil, errFromCode(cErr)
 	}
-	d.dev = dev
+
 	return dev, nil
 }
 
-// TODO(codingllama): Openning/closing the device for every operation seems
-// wasteful. Consider changing it.
-func (d *Device) close(dev *C.fido_dev_t) {
-	d.Lock()
-	d.dev = nil
-	d.Unlock()
+func (d *Device) getDevice() (*C.fido_dev_t, error) {
+	if d == nil {
+		return nil, errors.New("device nil")
+	}
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.dev == nil {
+		return nil, errors.New("device closed or not initialized")
+	}
+
+	return d.dev, nil
+}
+
+// Close closes and frees the underlying libfido2 device.
+// All devices must be explicitly closed. Closing a device multiple times is
+// harmless, as only the first Close takes effect.
+func (d *Device) Close() error {
+	if d == nil {
+		return nil
+	}
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
 
-	if cErr := C.fido_dev_close(dev); cErr != C.FIDO_OK {
-		logger.Errorf("%v", errors.Wrap(errFromCode(cErr), "failed to close"))
+	if d.dev == nil {
+		return nil
 	}
-	C.fido_dev_free(&dev)
+
+	cErr := C.fido_dev_close(d.dev)
+	C.fido_dev_free(&d.dev)
+	d.dev = nil // Likely already freed by fido_dev_free, nil again to be safe.
+	if cErr != C.FIDO_OK {
+		return fmt.Errorf("close device: %w", errFromCode(cErr))
+	}
+
+	return nil
 }
 
 // Cancel an action.
 func (d *Device) Cancel() error {
-	d.Lock()
-	defer d.Unlock()
-	if d.dev != nil {
-		if cErr := C.fido_dev_cancel(d.dev); cErr != C.FIDO_OK {
-			return errors.Wrap(errFromCode(cErr), "failed to cancel")
-		}
+	if d == nil {
```

Comment:
It's good to stay calm and not panic, but is there a valid use case where you call `Cancel()` on nil except for a mistake?

Reply:
It's probably a programming mistake if this happens, although not doing anything is reasonable for a cancel. Other methods should blow up.

As this is a library, I'd rather be safer than normal.